### DISCUSSION TITLE
Store sensitive configuration as encrypted data

### DISCUSSION THREAD

**Discussion by clotairer** [PrestaShop/PrestaShop#32238] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Some sensitive data like API key of payment, delivery, ... or native SMTP password are stored as a clear plain text in the database.

In case of SQL Injection exploit, a hacker can in chain access and update merchant values by corrupted data to divert a stream of payments, emails, employee accounts, ...

### Describe the solution you'd like

Configurations of PrestaShop are today very flexible and interesting to manage module configuration by language, shops, ...
It could be nice to store sensitive configuration as encrypted data.

### Additional context

I make a Pull Request as a POC.

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by touchweb-vincent** [comment]
And if i can add something : this is urgent.

**Comment by jf-viguier** [comment]
A must have

**Comment by matks** [comment]
Hello @clotairer that is a really nice idea and actually another prestashop user mentioned  it a few months ago 

We however ran into some problems: let's say we merge this PR in 8.1.0 . What happens for users who upgrade from 8.0.0 to 8.1.0 ? They start with an unencrypted database then after upgrade the database content is expected to be encrypted.

And we know Configuration mysql database can sometimes contain thousand of rows, so I'm afraid of performing a runtime encryption during the database upgrade 😅 .

We imagined a few ideas but we were not very satisfied with them:

1. Use a feature flag to enable the property: new shops have encryption ON, shops which do upgrade have encryption OFF. But how do people turn it on, that's a big question.

2. Add a SQL column `is_encrypted` in front of each configuration value ; and 'when user inputs a new configuration value or edits a configuration value ; encrypt it'. So the database content will slowly get encrypted as times goes by.

Another problem is how modules use configuration database today. Some may fetch data from it directly using SQL queries (and maybe use JOIN with other tables) so this will break this behavior. That's why maybe we should start with a few selected rows 😄 like SMTP credentials but not all of it.

Also, I think I remember @jocel1 telling us that Configuration database content is actually loaded _on every call to front-office_ maybe sometimes multiple times. This needs to be verified because my memory is vague, but then we need to check that we do not introduce a big performance issue.

So the idea is good but the path forward delivering it without issues will be long 🏰

**Comment by matks** [comment]
Note: I can remember discussing this almost 2 years ago https://github.com/PrestaShop/PrestaShop/issues/22540#issuecomment-753858590

**Comment by jocel1** [comment]
I confirm the Configuration table is loaded fully on every page calls into a php array. If you plan to implement this feature, I would recommand to implement a mecanism which would store the parsed result in a key/value database like redis for example, and handle properly the invalidation when the table is updated through a call to the Configuration::updateValue / ::set etc...

**Comment by jf-viguier** [comment]
The idea is to use encryption only on sensitive information (tokens, passwords), so IMO the perf is not a big concern for less than 10 configuration keys.

**Comment by clotairer** [comment]
I'm not sure it's important to have all data encrypted. Only secret API keys, ... I understand the performance issue of the current Configuration, but that's not this issue. Security is a very hot topic.

**Comment by jf-viguier** [comment]
Just read the PR : 
"In a module, just use CryptedConfiguration::updateValue('CONFIG_KEY', 'clear value'); to store a configuration and CryptedConfiguration::get('CONFIG_KEY') to get the clear value stored in the database."

**Comment by paulnoelcholot** [comment]
Hello @clotairer 

Thank you for your suggestion. The Maintainer Team will take it into consideration for future developments. 

Thanks for your PR :rocket: 

Thank you

**Comment by clotairer** [comment]
@paulnoelcholot my expectation exceeds just "consideration for future developments"...

**Comment by clotairer** [comment]
@matks is it useful if I work on my PR with none legacy classes and your "not very satisfied" suggestions ? Or you imagine only a solution with all or none configuration encrypted ?

**Comment by touchweb-vincent** [comment]
@clotairer For this evolution and according to our recent discovers, it's critical to think about the necessity to check that the token is really existing AND is not empty if a module need it expressly - even if the module is not installed (ykwim).

It would be smart to constrain module's developpers to declared theirs "sensitive/critical tokens" (which protect SU actions or similar) which MUST exist and NOT be empty - on the master class declaration of their modules ?

**Comment by eternoendless** [comment]
Wouldn't it be better to use environment variables instead of storing encrypted values in DB?

**Comment by clotairer** [comment]
Why do you mean exactly. For instance api credential of a payment module for instance PrestaShop Checkout with Paypal ? Where are stored API keys ?
Where are store credential of SMTP feature ?

**Comment by eternoendless** [comment]
Point taken. We'd still need to store the encryption key somewhere. What do you suggest?

**Reply by clotairer** [reply]
The easiest way in my mind as developper use an object like Configuration that store clear or encrypt configuration.
And with a polyfill to use the same object (or by default an object that have the same algorithm) in modules for PS 1.7 and PS 8.
My POC was here : #29677

**Reply by eternoendless** [reply]
Thing is, you don't want to store the encryption key in the same place where you store the encrypted content (i.e. the database), otherwise it amounts to keeping your house key in the keyhole ;)

**Reply by clotairer** [reply]
Off course, it use the new cookie key + a salt witch is the name of the configuration.

**Reply by jolelievre** [reply]
Although the secret manager proposition is a good idea, I think it'd be better if PrestaShop is able to handle this independently it would prevent a dependency on a third part service, no need to create an account and set credentials, no external call and it can be used and tested more easily locally

**Reply by KodeToad** [reply]
parameters.php is world readable.
```
# ls -l /var/www/prestashop/app/config/parameters.php
-rw-r--r-- 1 www-data www-data 3167 Oct  5 14:45 /var/www/prestashop/app/config/parameters.php
```

EDITED by @matks: snarky comments are useless and [against the Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).

**Comment by nicosomb** [comment]
@PrestaShop/tech-council Could you please give your opinion about this discussion please?

**Reply by eternoendless** [reply]
I think a good solution would be to encrypt values using an encryption key which is configured as an environment variable. I'm still unsure about whether the encryption/decryption should be provided by PrestaShop itself or let modules handle it and only provide the raw storage through Configuration.
